src/Cli/Microsoft.DotNet.Cli.Utils/MSBuildForwardingAppWithoutLogging.cs

```diff
@@ -34,26 +46,103 @@ internal class MSBuildForwardingAppWithoutLogging
 
         public MSBuildForwardingAppWithoutLogging(IEnumerable<string> argsToForward, string msbuildPath = null)
         {
-            _forwardingApp = new ForwardingAppImplementation(
-                msbuildPath ?? GetMSBuildExePath(),
-                _msbuildRequiredParameters.Concat(argsToForward.Select(Escape)),
-                environmentVariables: _msbuildRequiredEnvironmentVariables);
+            string defaultMSBuildPath = GetMSBuildExePath();
+
+            _argsToForward = argsToForward;
+            MSBuildPath = msbuildPath ?? defaultMSBuildPath;
+
+            // If DOTNET_CLI_EXEC_MSBUILD is set or we're asked to execute a non-default binary, call MSBuild out-of-proc.
+            if (AlwaysExecuteMSBuildOutOfProc || string.Compare(MSBuildPath, defaultMSBuildPath, StringComparison.OrdinalIgnoreCase) != 0)
+            {
+                _forwardingApp = new ForwardingAppImplementation(
+                    MSBuildPath,
+                    _msbuildRequiredParameters.Concat(argsToForward.Select(Escape)),
+                    environmentVariables: _msbuildRequiredEnvironmentVariables);
+            }
         }
 
         public virtual ProcessStartInfo GetProcessStartInfo()
```

Comment:
The code structure is indeed odd. I think that making `MSBuildForwardingApp` derive from `MSBuildForwardingAppWithoutLogging` would make much more sense and would naturally lead to what you're suggesting. I'll do it separately if you don't mind, so that this PR stays focused on the in-proc logic.